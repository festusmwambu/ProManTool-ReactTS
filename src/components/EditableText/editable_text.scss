.editable-text {
    cursor: pointer;
    display: flex;

    &__icon {
        opacity: 0;
        transition: .3s;
    }

    &:hover {
        & > * {
            text-decoration: underline;
        }

        .editable-text__icon {
            opacity: 1;
        }
    }
}