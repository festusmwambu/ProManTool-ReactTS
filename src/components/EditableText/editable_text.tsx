import { createElement, useState } from "react";
import useOnclickOutside from "react-cool-onclickoutside";
import { useForm } from "react-hook-form";
import cn from "classnames";

import "./editable_text.scss";



interface EditableTextProps {
    text: string | undefined;
    tag: "h1" | "h2" | "h3" | "h4" | "h5" | "h6" | "span" | "p";
    textClassName?: string;
    inputClassName?: string;
    onSuccess(value: string, callback: () => void): void;
    onCancel?(): void;
}

const EditableText = ({ text, tag, textClassName, inputClassName, onSuccess, onCancel, ...rest }: EditableTextProps) => {
    const { handleSubmit, register } = useForm<{ newValue: string }>(); // Adjust the type of the newValue here

    const [ editing, setEditing ] = useState(false);

    const handleSubmitSuccess = ({ newValue }: { newValue: string}) => {
        onSuccess(newValue, () => {
            setEditing(false);
        });
    };

    const cancel = () => {
        setEditing(false);

        if(onCancel) {
            onCancel();
        }
    };

    const clickOutsideRef = useOnclickOutside(cancel);

    return (
        <div {...rest} ref={clickOutsideRef} className="editable-text" onClick={() => setEditing(true)}>
            {!editing ? (
                createElement(
                    tag, 
                    { className: cn("editable-text__title", textClassName) }, 
                    (
                        <>
                            {text} <i className="fas fa-pencil-alt editable-text__icon" />
                        </>
                    )
                )
            ) : (
                <form onSubmit={handleSubmit(handleSubmitSuccess)}>
                    <input type="text" className={cn(inputClassName)} {...register("newValue", { required: "Required" })} defaultValue={text} />
                </form>
            )}
        </div>
    );
};

export default EditableText;