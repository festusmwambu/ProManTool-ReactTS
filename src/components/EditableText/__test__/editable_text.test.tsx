import { fireEvent, render, screen } from "@testing-library/react";

import EditableText from "../editable_text";



describe("<EditableText />", () => {
    it("Matches snapshot", () => {
        render(
            <EditableText text="test text" tag="p" onSuccess={jest.fn()} />
        );
        
        // Use screen.queryByTestId method
        const EditableTextElement = screen.queryByTestId("editable-text-element");
        expect(EditableTextElement).toMatchSnapshot();
    });

    it("Receives text", () => {
        const testText = "testText";

        render(
            <EditableText text={testText} tag="p" onSuccess={jest.fn()} />
        );
        
        const EditableTextElement = screen.queryByTestId("editable-text-element");
        expect(EditableTextElement).toHaveTextContent(testText);
    });

    it("Receives tag", () => {
        const testTag = "p";

        render(
            <EditableText text="testText" tag={testTag} onSuccess={jest.fn()} />
        );
        
        const EditableTextElement = screen.queryByTestId("editable-text-element");
        expect(EditableTextElement).toLowerCase().toBe(testTag);
    });

    it("Editable mode", () => {
        render(
            <EditableText text="testText" tag="p" onSuccess={jest.fn()} />
        );
        
        const EditableTextElement = screen.queryByTestId("editable-text-element");

        fireEvent.click(EditableTextElement as HTMLElement);

        expect(screen.getByRole("textbox")).toBeInTheDocument();
    });

    it("Input receives classname", () => {
        const inputClassName = "testClassName";

        render(
            <EditableText text="testText" tag="p" inputClassName={inputClassName} onSuccess={jest.fn()} />
        );
        
        const EditableTextElement = screen.queryByTestId("editable-text-element");

        fireEvent.click(EditableTextElement as HTMLElement);

        expect(screen.getByRole("textbox")).toHaveClass(inputClassName);
    });

    it("Input receives default text", () => {
        const textValue = "Some test value";

        render(
            <EditableText text={textValue} tag="p" onSuccess={jest.fn()} />
        );
        
        const EditableTextElement = screen.queryByTestId("editable-text-element");

        fireEvent.click(EditableTextElement as HTMLElement);
        
        expect(screen.getByRole("textbox").getAttribute("value")).toBe(textValue);
    });

    it("onSuccess called", async () => {
        const onSuccess = jest.fn();

        render(
            <EditableText text="testtext" tag="p" onSuccess={onSuccess} />
        );
        
        const EditableTextElement = screen.queryByTestId("editable-text-element");
        
        fireEvent.click(EditableTextElement as HTMLElement);

        fireEvent.submit(screen.getByRole("form"));

        expect(onSuccess).toBeCalledTimes(1);
    });
});