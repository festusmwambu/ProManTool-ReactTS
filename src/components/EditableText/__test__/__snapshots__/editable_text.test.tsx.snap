// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableText /> Matches snapshot 1`] = `
<div
  className="editable-text"
>
  <p
    className="editable-text__title"
  >
    test text
     
    <i
      className="fas fa-pencil-alt editable-text__icon"
    />
  </p>
</div>
`;